src/dawn/Optimizer/PassFieldVersioning.cpp

```diff
@@ -89,71 +91,165 @@ static void reportRaceCondition(const Statement& statement,
   }
 }
 
-} // anonymous namespace
+static void addAssignmentToDoMethod(std::unique_ptr<iir::DoMethod>& domethod, int assignmentID,
+                                    int assigneeID,
+                                    const std::shared_ptr<iir::StencilInstantiation>& context,
+                                    boost::optional<iir::Extents>& readExtents) {
+  // Create the StatementAccessPair of the assignment with the new and old variables
+  auto fa_assignee =
+      std::make_shared<FieldAccessExpr>(context->getMetaData().getNameFromAccessID(assigneeID));
```

Comment:
This is the case for every one of the >25 calls to `insertExprToAccessID`. Is this functionality generally flawed?

Reply:
not always, isnt? In this example
https://github.com/MeteoSwiss-APN/dawn/blob/master/src/dawn/Optimizer/StatementMapper.cpp#L348
we mapped an existing expr, we dont create it. 
I think we would need the two levels, but for cases where we create the shared_ptr<Expr> and insert the mapping it would make sense to have it in a maker. It follows the ideas of the last big refactoring to always have the IIR and metadata in consistent state, we just didnt do it yet. 

Actually generalizing the idea, instead of a makeFieldAccessExpr, we could have a 
```
template<typename ExprType, typename ... Args>
makeExpr(Args && ... args)
```

Reply:
sorry, I was unclear. What I meant is that we enter this state of inconsistency whenever we create a new Expr / Stmt before we do the mapping. This happens in various parts of the code.

I think the generalization makes sense and that way we can have a register function that handles all the things that need to be handled (filling this map, filling `accessIfTpye`, potential future thinigs) in a centralized place. I would do this in a separate PR though. Does that make sense?